package Genymotion;

message Status {
  enum Code {
    Ok                = 0;
    GenericError      = 1;
    InvalidRequest    = 2;
    OkWithInformation = 3;
    NotImplemented    = 4;
  }

  required Code code = 2;
  optional string description = 3 [default = ""];
}

message Value {
  enum Type {
    Bool   = 1;
    Int    = 2;
    Float  = 3;
    String = 4;
    Bytes  = 5;
  }

   required Type    type        = 1;

   optional bool    boolValue   = 2;
   optional int32   intValue    = 3;
   optional float   floatValue  = 4;
   optional string  stringValue = 5;
   optional bytes   bytesValue  = 6;
}


message Parameter {
 enum Type {
    None               = 0;
    AndroidVersion     = 1;  // string
    BuildNumber        = 2;  // string
    BatteryMode        = 3;
    BatteryLevel       = 4;  // uint in [0; 100]
    BatteryStatus      = 5;  // "Charging", "Discharging", "Not charging", "Full"
    GpsStatus          = 6;  // boolean for Enabled/Disabled
    GpsLatitude        = 7;  // degrees: float in [-90; 90[
    GpsLongitude       = 8;  // degrees: float in [-180; 180[
    GpsAltitude        = 9;  // meters:  float in [-10000; 10000]
    GpsAccuracy        = 10; // meters:  float in [0; 200]
    GpsBearing         = 11; // degrees: float in [0; 360[
    AccemerometerX     = 12; // acceleration: float in m/s^2
    AccemerometerY     = 13; // acceleration: float in m/s^2
    AccemerometerZ     = 14; // acceleration: float in m/s^2
  }

  required Type  type = 1;
  optional Value value = 2;
}

message Request {
  enum Type {
    None            = 0;
    Ping            = 1;
    GetParam        = 2;
    SetParam        = 3;
  }

  required  Type       type        = 1;
  optional  Parameter  parameter   = 2;
}

message Reply {
  enum Type {
    None   = 0;
    Error  = 1;
    Pong   = 2;
    Value  = 3;
  }

  required  Type    type    = 1;
  required  Status  status  = 2;
  optional  Value   value   = 3;
}
